### PR TITLE
chore: Metrics changes for Mvtx UDF, on-success sink and drop metrics

### PR DESCRIPTION
**Pull request by vaibhavtiwari33**
### What this PR does / why we need it

The PR aims to fix the following issues:
* Add metrics for whenever upstream components mark messages to be dropped in sink.
* Add metrics for monovertex udf/mapper
* Add metrics for on-success sink.

### Testing
How was this tested (unit/integration/manual)? Include commands, links, or screenshots if applicable.

### Special notes for reviewers
Anything notable for review (risk, rollout, follow-ups).

### CODE REVIEW

**Comment by vigith** [comment]
> Patch coverage is 67.37288% with 77 lines in your changes missing coverage. Please review.

nothing we can do?

**Comment by vaibhavtiwari33** [comment]
> > Patch coverage is 67.37288% with 77 lines in your changes missing coverage. Please review.
> 
> nothing we can do?

My bad, looks like the condition for generating mvtx specific metrics might be causing this. Let me add tests for the same.

File: docs/operations/metrics/metrics.md [hunk 1]
```diff
@@ -22,6 +22,8 @@ These metrics can be used to determine throughput of your pipeline.
...
 | `forwarder_fbsink_write_total`             | Counter     | `pipeline=<pipeline-name>` <br> `vertex=<vertex-name>` <br> `vertex_type=<vertex-type>` <br> `replica=<replica-index>` <br> `partition_name=<partition-name>` | Provides the total number of messages written to a fallback sink                                                |
 | `forwarder_fbsink_write_bytes_total`       | Counter     | `pipeline=<pipeline-name>` <br> `vertex=<vertex-name>` <br> `vertex_type=<vertex-type>` <br> `replica=<replica-index>` <br> `partition_name=<partition-name>` | Provides the total number of bytes written to a fallback sink                                                   |
+| `forwarder_onssink_write_total`            | Counter     | `pipeline=<pipeline-name>` <br> `vertex=<vertex-name>` <br> `vertex_type=<vertex-type>` <br> `replica=<replica-index>` <br> `partition_name=<partition-name>` | Provides the total number of messages written to a on-success sink                                              |
```
Comment (adarsh0728) [inline docs/operations/metrics/metrics.md]: should we change the name? `forwarder_onsuccess_sink_write_total` or something informative so that it is easier for users to understand.

File: docs/operations/metrics/metrics.md [hunk 2]
```diff
@@ -11,7 +11,7 @@ These metrics in combination can be used to determine the overall health of your
 These metrics can be used to determine throughput of your pipeline.
```
Comment (adarsh0728) [inline docs/operations/metrics/metrics.md:11]: one general suggestion - can we have following headings instead of Traffic, Latency, Errors etc

- Pipeline Metrics
- MonoVertex Metrics
- ISB Metrics
- Controller Metrics
- SDK Info Metrics

WDYT?  Can be taken up in follow up PR as well.

Comment (vaibhavtiwari33) [reply docs/operations/metrics/metrics.md:11]: That would be easier to navigate, we can have sub-sections within each as required.

File: docs/operations/metrics/metrics.md (hunk 1 above)
Comment (vaibhavtiwari33) [reply docs/operations/metrics/metrics.md]: Sure, yes. I went with this format since the on success sink is named `ons-sink`
For metrics purposes, I think it should be fine to have long names.

**Comment by vaibhavtiwari33** [comment]
> no need to test the metrics per se, but we should add negative tests that invoke drop behavior?

Added a test that verifies drop behavior in [this PR](https://github.com/numaproj/numaflow/pull/3134)

**Comment by vaibhavtiwari33** [comment]
> > Patch coverage is 67.37288% with 77 lines in your changes missing coverage. Please review.
> 
> nothing we can do?

Tried adding test for this, but seems like setting env variable would require unsafe block

**Comment by vigith** [comment]
let's also address the comments in https://github.com/numaproj/numaflow/issues/3136#issuecomment-3776627056 and https://github.com/numaproj/numaflow/issues/3136#issuecomment-3776659273 in this PR

**Comment by vigith** [comment]
> > > Patch coverage is 67.37288% with 77 lines in your changes missing coverage. Please review.
> > 
> > 
> > nothing we can do?
> 
> Tried adding test for this, but seems like setting env variable would require unsafe block

the trick is to move `env` one level up and make the method accept env as a `HashMap`. This is what we do in config, etc.

File: docs/operations/metrics/metrics.md (hunk 1 above)
Comment (vaibhavtiwari33) [reply docs/operations/metrics/metrics.md]: done

File: rust/numaflow-core/src/sinker/sink.rs [hunk 3]
```diff
@@ -603,6 +651,33 @@ impl SinkWriter {
     }
 }
 
+/// Sends count of messages marked for explicit drop by the user
+/// Currently pub(crate) to allow usage by the bypass_router.
+/// TODO: Should this be a separate metric? Currently reason is different for pipeline
+pub(crate) fn send_drop_metrics(is_mono_vertex: bool, dropped_messages_count: usize) {
+    if dropped_messages_count > 0 {
```
Comment (vigith) [inline rust/numaflow-core/src/sinker/sink.rs]: NIT: I don't think you need this check.. `incr_by(0)` is valid and code will look cleaner.

File: rust/numaflow-core/Cargo.toml [hunk 4]
```diff
@@ -74,5 +74,6 @@ aws-smithy-mocks = "0.1.1"
 k8s-openapi = { version = "0.24.0", features = ["v1_32"] }
 numaflow-kafka = { workspace = true, features = ["kafka-tests-utils"] }
 numaflow-pulsar = { workspace = true, features = ["pulsar-tests-utils"] }
+serial_test = "3.3.1"
```
Comment (vigith) [inline rust/numaflow-core/Cargo.toml:77]: the reason you need this is to make sure tests which require metrics update should run serially since metrics is a global state, correct?

**Comment by vigith** [comment]
I did not review `metrics.md` :)

File: rust/numaflow-core/Cargo.toml (hunk 4 above)
Comment (vaibhavtiwari33) [reply rust/numaflow-core/Cargo.toml:77]: Yes, that's correct

File: rust/numaflow-core/src/sinker/sink.rs (hunk 3 above)
Comment (vaibhavtiwari33) [reply rust/numaflow-core/src/sinker/sink.rs]: That's a good point actually, let me address this.

**Comment by vaibhavtiwari33** [comment]
Following is the updated metrics.md that relate to updating numaflow version in the exposition format section:


Screenshot 2026-01-21 at 7 09 28 PM

File: docs/operations/metrics/metrics.md (hunk 2 above)
Comment (vaibhavtiwari33) [reply docs/operations/metrics/metrics.md:11]: Created a separate issue for this

File: rust/numaflow-core/src/sinker/sink.rs (hunk 3 above)
Comment (vaibhavtiwari33) [reply rust/numaflow-core/src/sinker/sink.rs]: Updated
